src/main/java/org/elasticsearch/cluster/DiffableUtils.java

```diff
@@ -0,0 +1,204 @@
+     */
+    public static <T extends Diffable<T>> Diff diff(ImmutableOpenMap<String, T> before, ImmutableOpenMap<String, T> after) {
+        assert after != null && before != null;
+        Map<String, Diff> diffs = newHashMap();
+        List<String> deletes = newArrayList();
+        Map<String, T> adds = newHashMap();
+        for (ObjectCursor<String> key : before.keys()) {
+            if (!after.containsKey(key.value)) {
+                deletes.add(key.value);
+            }
+        }
+        for (ObjectObjectCursor<String, T> partIter : after) {
+            T beforePart = before.get(partIter.key);
+            if (!partIter.value.equals(beforePart)) {
+                if (beforePart == null) {
+                    adds.put(partIter.key, partIter.value);
+                } else {
+                    diffs.put(partIter.key, partIter.value.diffs(beforePart));
+                }
+            }
+        }
+        return new CompositeDiff<>(deletes, diffs, adds);
+    }
+
+    /**
+     * Calculates diffs between two ImmutableMaps of Diffable objects
+     */
+    public static <T extends Diffable<T>> Diff diff(ImmutableMap<String, T> before, ImmutableMap<String, T> after) {
+        assert after != null && before != null;
+        Map<String, Diff> diffs = newHashMap();
+        List<String> deletes = newArrayList();
+        Map<String, T> adds = newHashMap();
+        for (String key : before.keySet()) {
+            if (!after.containsKey(key)) {
+                deletes.add(key);
+            }
+        }
+        for (ImmutableMap.Entry<String, T> partIter : after.entrySet()) {
+            T beforePart = before.get(partIter.getKey());
+            if (!partIter.getValue().equals(beforePart)) {
+                if (beforePart == null) {
+                    adds.put(partIter.getKey(), partIter.getValue());
+                } else {
+                    diffs.put(partIter.getKey(), partIter.getValue().diffs(beforePart));
+                }
+            }
+        }
+        return new CompositeDiff<>(deletes, diffs, adds);
+    }
+
+
+    /**
+     * Reads Diff object produced by {@link org.elasticsearch.cluster.DiffableUtils#diff(org.elasticsearch.common.collect.ImmutableOpenMap, org.elasticsearch.common.collect.ImmutableOpenMap)} method
+     *
+     * The diffs are applied to the before object.
+     */
+    public static <T extends Diffable<T>> ImmutableOpenMap<String, T> readDiff(ImmutableOpenMap<String, T> before, StreamInput in, final T proto) throws IOException {
+        return readDiff(before, in, new KeyedReader<T>() {
+            @Override
+            public T readFrom(StreamInput in, String key) throws IOException {
+                return proto.readFrom(in);
+            }
+        });
+    }
+
+    /**
+     * Reads Diff object produced by {@link org.elasticsearch.cluster.DiffableUtils#diff(org.elasticsearch.common.collect.ImmutableOpenMap, org.elasticsearch.common.collect.ImmutableOpenMap)} method
+     *
```

Comment:
can we also add a line saying this variant allows for custom value deserialization based on the map using a KeyedReader?  I got confused by why this is useful (until I saw how it is used).